### PR TITLE
Type-annotate and unit-test timeformat, fix off-by-one day in parseShortDate()

### PR DESCRIPTION
**Pull request by martinpitt**
This is a prerequisite for #20653 and also fixes an important bug.

### CODE REVIEW

**Comment by allisonkarlitskaya** [comment]
> I completed the unit tests, and hold and below, they found a bug! Fix added.

I think you mean [lo and behold](https://en.wiktionary.org/wiki/lo_and_behold) :)

**Comment by martinpitt** [comment]
This unfortunately "affects" just about every test, so this exposes all our flakes. /me leans on retry button

File: pkg/lib/timeformat.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+type Time = Date | number;
+
+// general Intl.DateTimeFormat formatter object
+export const formatter = (options?: DTFOptions) => new Intl.DateTimeFormat(dateFormatLang(), options);
+
+// common formatters; try to use these as much as possible, for UI consistency
+// 07:41 AM
+export const time = (t: Time): string => formatter({ timeStyle: "short" }).format(t);
+// 7:41:26 AM
+export const timeSeconds = (t: Time): string => formatter({ timeStyle: "medium" }).format(t);
```
Comment (cockpituous) [inline pkg/lib/timeformat.ts:34]: This added line is not executed by any test.

File: pkg/lib/timeformat.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ * based on https://github.com/unicode-cldr/cldr-core/blob/master/supplemental/weekData.json#L59
+ * However, we don't have translations for most locales, and cockpit.language does not even contain
+ * the country in most cases, so this is just an approximation.
+ * Most locales start the week on Monday (day 1), so default to that and enumerate the others.
+ */
+
+const first_dow_sun = ['en', 'ja', 'ko', 'pt', 'pt_BR', 'sv', 'zh_CN', 'zh_TW'];
+
+export function firstDayOfWeek(): number {
+    return first_dow_sun.indexOf(cockpit.language) >= 0 ? 0 : 1;
```
Comment (cockpituous) [inline pkg/lib/timeformat.ts:76]: This added line is not executed by any test.

File: pkg/lib/cockpit-components-shutdown.jsx [hunk 3]
```diff
@@ -31,7 +31,7 @@ import { DatePicker } from "@patternfly/react-core/dist/esm/components/DatePicke
 import { TimePicker } from "@patternfly/react-core/dist/esm/components/TimePicker/index.js";
 
 import { ServerTime } from 'serverTime.js';
-import * as timeformat from "timeformat.js";
+import * as timeformat from "timeformat";
```
Comment (allisonkarlitskaya) [inline pkg/lib/cockpit-components-shutdown.jsx:34]: We added a rule to Cockpit Files to forbid file extension on imports.....

Topic for another PR :)

File: pkg/lib/timeformat.ts [hunk 4]
```diff
@@ -8,47 +8,60 @@ import { parse, formatDistanceToNow } from 'date-fns';
 import * as locales from 'date-fns/locale';
 
 // this needs to be dynamic, as some pages don't initialize cockpit.language right away
-export const dateFormatLang = () => cockpit.language.replace('_', '-');
+export const dateFormatLang = (): string => cockpit.language.replace('_', '-');
```
Comment (allisonkarlitskaya) [inline pkg/lib/timeformat.ts:11]: Was it necessary to annotate the return type?

In general, return types can be omitted, and I usually do that.  On the other hand, I've seen noises from "smart people" (including the TypeScript team) which suggest that type inference in other parts of the program will be faster if you explicitly annotate...

This isn't a request for change — just random musings.

File: pkg/lib/timeformat.ts [hunk 5]
```diff
@@ -8,47 +8,60 @@ import { parse, formatDistanceToNow } from 'date-fns';
...
+const dateFormatLangDateFns = (): string => {
     if (cockpit.language == "en") return "enUS";
     else return cockpit.language.replace('_', '');
 };
 
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const dateFnsLocale = (): any => (locales as any)[dateFormatLangDateFns()];
+
+type DTFOptions = {
+    [key: string]: string;
```
Comment (allisonkarlitskaya) [inline pkg/lib/timeformat.ts:22]: I think that's spelled `Intl.DateTimeFormatOptions`.

In general it's helpful to go looking inside of `node_modules/typescript/lib/` or `node_modules/@types` to see what this thing is "officially" called, and use that.

File: pkg/lib/timeformat.ts [hunk 6]
```diff
@@ -8,47 +8,60 @@ import { parse, formatDistanceToNow } from 'date-fns';
...
+export const dateFormatLang = (): string => cockpit.language.replace('_', '-');
 
-const dateFormatLangDateFns = () => {
+const dateFormatLangDateFns = (): string => {
     if (cockpit.language == "en") return "enUS";
     else return cockpit.language.replace('_', '');
 };
 
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const dateFnsLocale = (): any => (locales as any)[dateFormatLangDateFns()];
```
Comment (allisonkarlitskaya) [inline pkg/lib/timeformat.ts:19]: naja... I guess that part is going away soon...

File: pkg/lib/timeformat.ts (hunk 1 above)
Comment (allisonkarlitskaya) [reply pkg/lib/timeformat.ts:34]: I demand coverage!!

Comment (allisonkarlitskaya) [reply pkg/lib/timeformat.ts:34]: ... or the :axe:

File: pkg/base1/test-timeformat.ts [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    cockpit.language = "de";
+    assert.equal(timeformat.distanceToNow(seconds), "weniger als 1 Minute");
+    assert.equal(timeformat.distanceToNow(seconds, true), "vor weniger als 1 Minute");
+
+    assert.equal(timeformat.distanceToNow(now - 4 * DAY), "4 Tage");
+    assert.equal(timeformat.distanceToNow(now - 21 * DAY), "21 Tage");
+    assert.equal(timeformat.distanceToNow(now - 62 * DAY), "2 Monate");
+});
+
+QUnit.test("firstDayOfWeek", assert => {
```
Comment (allisonkarlitskaya) [inline pkg/base1/test-timeformat.ts:120]: ah right.  It would be nice if our coverage reporting also covered the QUnit runs...

**Review by allisonkarlitskaya — CHANGES_REQUESTED** [review]
I'm personally insulted that you didn't choose `America/Toronto` in your manual testing.

Is there a way we can whack the TZ of the browser from inside so that we can get those cases under regression tests?  Nobody is ever going to run that command ever again.

Otherwise, looks good except for the egregious use of an indexed access type.

File: pkg/lib/timeformat.ts (hunk 4 above)
Comment (martinpitt) [reply pkg/lib/timeformat.ts:11]: I don't think it was *necessary* -- but if we do this, IMHO we should explicitly type all external API, to make it more self-documenting and avoid ambiguity. I like the python mypy approach to require types in declarations, even if inference could figure them out.

File: pkg/lib/timeformat.ts (hunk 1 above)
Comment (martinpitt) [reply pkg/lib/timeformat.ts:34]:
```
pkg/metrics/metrics.jsx:        let tooltip = timeformat.timeSeconds(time) + "\n\n";
```
we don't test the tooltip, but the qunit test does cover this.

File: pkg/lib/timeformat.ts (hunk 2 above)
Comment (martinpitt) [reply pkg/lib/timeformat.ts:76]: that line is covered, just not both cases -- we dont' test the shutdown time or timer create dialog in non-English. But again, qunit test covers this.

File: pkg/lib/timeformat.ts (hunk 5 above)
Comment (martinpitt) [reply pkg/lib/timeformat.ts:22]: Thanks! Fixed.

**Comment by martinpitt** [comment]
> Is there a way we can whack the TZ of the browser from inside so that we can get those cases under regression tests? 

Not from within test-timeformat.ts,  as they all run in the same browser instance. There's a [CDP call](https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setTimezoneOverride), but we don't have access to that from inside a QUnit test.

A possible entry point would be in test/pytest/test_browser.py , it could run the timeformat test in separate test-server/browser runs. It's a bit ad-hoc, but I agree it'd be worth it. I did that now, and it reproduces the bug very nicely (and confirms the fix). And yes, it uses A/Toronto now :grin:

**Comment by martinpitt** [comment]
Ah bummer -- we [don't have timezones in Debian](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20655-fbdda0e9-20240626-072629-debian-testing-other/log.html), these can't run in the build environment. But that shouldn't run the browser pytest in the first place, we need a similar skip as `test_browser`.

File: test/pytest/test_browser.py [hunk 8]
```diff
@@ -46,3 +46,22 @@ def test_browser(html):
     subprocess.run(['test/common/tap-cdp', f'{BUILDDIR}/test-server',
                     sys.executable, '-m', *coverage, 'cockpit.bridge', '--debug',
                     f'./qunit/{html}'], check=True, stderr=subprocess.STDOUT)
+
+
+# run test-timeformat.ts in different time zones: west/UTC/east
+@pytest.mark.parametrize('tz', ['America/Toronto', 'Europe/London', 'UTC', 'Europe/Berlin', 'Australia/Sydney'])
```
Comment (allisonkarlitskaya) [inline test/pytest/test_browser.py:52]: Cool.

`test-timeformat` is also going to run from the normal `test_browser()` path, but I guess that's good: then we get the version without explicit `TZ` set.  It also means that `pytest -k browser` will still hit it (at least for the one case) but won't get bogged down with having to do the exhausting checking.

File: pkg/lib/timeformat.ts (hunk 4 above)
Comment (allisonkarlitskaya) [reply pkg/lib/timeformat.ts:11]: I'm going to start writing explicit return types from now on.

File: pkg/lib/timeformat.ts [hunk 9]
```diff
@@ -8,47 +8,56 @@ import { parse, formatDistanceToNow } from 'date-fns';
...
+
+type Time = Date | number;
+
 // general Intl.DateTimeFormat formatter object
-export const formatter = options => new Intl.DateTimeFormat(dateFormatLang(), options);
+export const formatter = (options?: Intl.DateTimeFormatOptions) => new Intl.DateTimeFormat(dateFormatLang(), options);
```
Comment (allisonkarlitskaya) [inline pkg/lib/timeformat.ts:24]: Much better, thanks!

**Review by allisonkarlitskaya — APPROVED** [review]
Nice little piece of work.  Thanks!

File: test/pytest/test_browser.py (hunk 8 above)
Comment (martinpitt) [reply test/pytest/test_browser.py:52]: Right, I deliberately didn't ignore it from `test_browser()`, it's also nice to test in your current time zone.

File: pkg/lib/timeformat.ts [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+type Time = Date | number;
+
+// general Intl.DateTimeFormat formatter object
+export const formatter = (options?: Intl.DateTimeFormatOptions) => new Intl.DateTimeFormat(dateFormatLang(), options);
+
+// common formatters; try to use these as much as possible, for UI consistency
+// 07:41 AM
+export const time = (t: Time): string => formatter({ timeStyle: "short" }).format(t);
+// 7:41:26 AM
+export const timeSeconds = (t: Time): string => formatter({ timeStyle: "medium" }).format(t);
```
Comment (cockpituous) [inline pkg/lib/timeformat.ts:30]: This added line is not executed by any test.

File: pkg/lib/timeformat.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+ * based on https://github.com/unicode-cldr/cldr-core/blob/master/supplemental/weekData.json#L59
+ * However, we don't have translations for most locales, and cockpit.language does not even contain
+ * the country in most cases, so this is just an approximation.
+ * Most locales start the week on Monday (day 1), so default to that and enumerate the others.
+ */
+
+const first_dow_sun = ['en', 'ja', 'ko', 'pt', 'pt_BR', 'sv', 'zh_CN', 'zh_TW'];
+
+export function firstDayOfWeek(): number {
+    return first_dow_sun.indexOf(cockpit.language) >= 0 ? 0 : 1;
```
Comment (cockpituous) [inline pkg/lib/timeformat.ts:72]: This added line is not executed by any test.

**Pull request merged**